streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -299,7 +318,17 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
                             restoreRecords.add(recordConverter.convert(record));
                         }
                     }
-                    offset = globalConsumer.position(topicPartition);
+                    try {
+                        offset = globalConsumer.position(topicPartition);
+                    } catch (final TimeoutException error) {
+                        // the `globalConsumer.position()` call should never block, because we know that we did
+                        // a successful `position()` call above for the requested partition and thus the consumer
+                        // should have a valid local position that it can return immediately
+
+                        // hence, a `TimeoutException` indicates a bug and thus we rethrow it as fatal `IllegalStateException`
+                        throw new IllegalStateException(error);
+                    }
+
                     stateRestoreAdapter.restoreBatch(restoreRecords);
                     stateRestoreListener.onBatchRestored(topicPartition, storeName, offset, restoreRecords.size());
                     restoreCount += restoreRecords.size();
```

Comment:
@vvcephei I actually had a follow up thought: given that we fetch data for a single partition only, should we trigger a timeout within this loop if `records` is empty (ie, `poll() did not return anything)? Otherwise, this restore loop might "hang" forever if we lose the connection to the broker -- IMHO, `task.timeout.ms` should cover this case? Otherwise, we block on startup as the initial global store loading would never finish and we don't even start the actual `StreamThreads`.

Reply:
Huh, interesting thought. Just to be clear, it looks like it would already block forever in this case today, right?
Yeah, it does seem like we should implement a similar non-progress timeout for this loop in that case.

Reply:
Yes, it would block forever today. Curious to hear what @guozhangwang thinks.

Reply:
My understanding is that for non-global state stores, we would also start ticking if we cannot make progress either due to exceptions or `poll()` returned no data, is that right? If yes, I'm +1 on covering the same for global state store here.